File: ios/stereoCamera/stereoCamera/Comm/Commands/SendGravity.swift
//
//  SendGravity.swift
//  stereoCamera
//
//  Created by hallmarklabs on 7/13/18.
//  Copyright © 2018 cody. All rights reserved.
//

import Foundation

class SendGravity : Command
{
    var gravity = Gravity(x: 0, y: 0, z: 0)
    
    override init()
    {
        super.init()
        doInit()
    }
    
    init(gravity:Gravity)
    {
        super.init()
        
        self.gravity = gravity
        doInit()
    }
    
    func doInit()
    {
        self.cmdtype = CommandTypes.RECEIVE_GRAVITY
        self.expectsResponse = false
    }
    
    override func send(comm: Comm) -> Bool
    {
        let result = super.send(comm: comm)
        if (!result) { return false }
        
        var bytes:[UInt8] = []
        bytes += Bytes.toByteArray(gravity.x)
        bytes += Bytes.toByteArray(gravity.y)
        bytes += Bytes.toByteArray(gravity.z)
        
        let sz = comm.write(buf: bytes)
        return (sz <= 0) ? false : true
    }
    
    override func receive(comm: Comm) -> Bool
    {
        let result = super.receive(comm: comm)
        if (!result) { return false }
        
        let bytes = [UInt8].init(repeating: 0, count: 4)
        
        let sz1 = comm.read(buffer: bytes)
        if (sz1 <= 0) { return false }
        gravity.x = Bytes.fromByteArray( bytes)
        
        let sz2 = comm.read(buffer: bytes)
        if (sz2 <= 0) { return false }
        gravity.y = Bytes.fromByteArray(bytes)
        
        let sz3 = comm.read(buffer: bytes)
        if (sz3 <= 0) { return false }
        gravity.z = Bytes.fromByteArray(bytes)
        
        
        return true
    }
}

struct Gravity
{
    var x:Float
    var y:Float
    var z:Float
}

File: android/app/src/main/java/com/munger/stereocamera/ip/command/commands/SendGravity.java
package com.munger.stereocamera.ip.command.commands;

import com.munger.stereocamera.ip.command.Comm;
import com.munger.stereocamera.ip.command.Command;

import java.io.IOException;

public class SendGravity extends Command
{
	public static class Gravity
	{
		public float x;
		public float y;
		public float z;
	}

	public Gravity gravity;

	public SendGravity()
	{
		super();
		doInit();
	}

	public SendGravity(Gravity gravity)
	{
		super();
		this.gravity = gravity;
		doInit();
	}

	private void doInit()
	{
		cmdtype = Type.RECEIVE_GRAVITY;
		expectsResponse = false;
	}

	@Override
	public boolean send(Comm comm)
	{
		boolean result = super.send(comm);
		if (!result)
			{ return false; }

		try
		{
			comm.putFloat(gravity.x);
			comm.putFloat(gravity.y);
			comm.putFloat(gravity.z);
		}
		catch(IOException e){
			return false;
		}

		return true;
	}

	@Override
	public boolean receive(Comm comm)
	{
		boolean result = super.receive(comm);
		if (!result)
			{ return false; }

		try
		{
			gravity = new Gravity();
			gravity.x = comm.getFloat();
			gravity.y = comm.getFloat();
			gravity.z = comm.getFloat();
		}
		catch(IOException e){
			return false;
		}

		return true;
	}
}
